Wikipedia/Code/WMFDatabaseHouseKeeper.swift

```diff
@@ -0,0 +1,113 @@
+import Foundation
+
+@objc class WMFDatabaseHouseKeeper : NSObject {
+    
+    @objc func performHouseKeepingOnManagedObjectContext(_ moc: NSManagedObjectContext)  throws {
+        
+        try deleteStaleUnreferencedArticles(moc)
+    }
+    
+    private func deleteStaleUnreferencedArticles(_ moc: NSManagedObjectContext) throws {
+        
+        /**
+ 
+        Find `WMFContentGroup`s more than 30 days old.
+ 
+        */
+        
+        let midnightTodayUTC = (Date() as NSDate).wmf_midnightUTCDateFromLocal!
+        let utcCalendar = NSCalendar.wmf_utcGregorian() as Calendar
+        let thirtyDaysAgoMidnightUTC = utcCalendar.date(byAdding: .day, value: -30, to: midnightTodayUTC, wrappingComponents: true)
+        
+        let allContentGroupFetchRequest = WMFContentGroup.fetchRequest()
+        allContentGroupFetchRequest.propertiesToFetch = ["key", "midnightUTCDate", "content", "articleURLString"]
+        
+        let allContentGroups = try moc.fetch(allContentGroupFetchRequest)
+        var referencedArticleKeys = Set<String>(minimumCapacity: allContentGroups.count * 5 + 1)
+        
+        for group in allContentGroups {
+            if group.midnightUTCDate?.compare(thirtyDaysAgoMidnightUTC!) == .orderedAscending {
+                moc.delete(group)
+                continue
+            }
+            
+            if let articleURLDatabaseKey = (group.articleURL as NSURL?)?.wmf_articleDatabaseKey {
+                referencedArticleKeys.insert(articleURLDatabaseKey)
+            }
+            
+            guard let content = group.content as Array! else {
```

Comment:
This line was a little weird to translate from ObjC. Actually I think we're ok just unwrapping the optional instead of the explicit cast because it's already defined as `[NSCoding]?`